lib/internal/url.js

```diff
@@ -0,0 +1,328 @@
+'use strict';
+
+function getPunycode() {
+  try {
+    return process.binding('icu');
+  } catch (err) {
+    return require('punycode');
+  }
+}
+const punycode = getPunycode();
+const binding = process.binding('url');
+const context = Symbol('context');
+const cannotBeBase = Symbol('cannot-be-base');
+
+function StorageObject() {}
+StorageObject.prototype = Object.create(null);
+
+class URL {
+  constructor(input, base) {
```

Comment:
Yeah, the js side of this is somewhat temporary at the moment while I focus on the parsing bits. Currently this is a close-enough-to-be-somewhat-functional impl of the IDL bits ;-). This is definitely helpful tho and I've added it to my list of todo's.